### PR TITLE
Enabling concat fast path for channels last inputs

### PR DESCRIPTION
**Pull request by kimishpatel**
Summary:
Updates concat kernel for contiguous input to support channels_last contig tensors.

This was tried on squeezenet model on pixel-2 device. It improves model perf by about 25%.

Test Plan:
test_cat_in_channels_last

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Comment by kimishpatel** [comment]
Note that, I think this can be combined with the exist kernel in CatKernel.cpp. However it would be good to parallelize that one. The original motivation for that one was to fix a perf regression but it did not have to be serial I believe.

**Comment by zou3519** [comment]
I don't know much about memory format so I am removing myself from the reviewers list. @VitalyFedyunin could you take a look at this?

**Comment by kimishpatel** [comment]
> I don't know much about memory format so I am removing myself from the reviewers list. @VitalyFedyunin could you take a look at this?

Ya no worries. I added you because gh showed me you reviewed those files recently. I will let other reviewers chime in.

**Comment by kimishpatel** [comment]
@VitalyFedyunin, can you please take a look at this? Thanks.

> Note that, I think this can be combined with the exist kernel in CatKernel.cpp. However it would be good to parallelize that one. The original motivation for that one was to fix a perf regression but it did not have to be serial I believe.
> I will try that out as well.

I tried parallelizing the serial kernel. The perf results were inconclusive and variance very high in some cases. For now just leaving that as is.

**Comment by kimishpatel** [comment]
@VitalyFedyunin, can you please take a look at this?

**Comment by kimishpatel** [comment]
@colesbury, can you help review this PR? Thanks.

File: test/test_torch.py [hunk 1]
```diff
@@ -7045,6 +7045,15 @@ def test_cat_out_channels_last(self, device):
...
         self.assertEqual(res1, res2)
 
+    def test_cat_in_channels_last(self, device):
+        x = torch.randn((4, 15, 8, 8))
+        y = torch.randn(x.shape)
+        res1 = torch.cat((x, y), dim=1)
+        x = x.clone().contiguous(memory_format=torch.channels_last)
+        y = y.clone().contiguous(memory_format=torch.channels_last)
+        res2 = torch.cat((x, y), dim=1)
+        self.assertEqual(res1, res2)
```
Comment (VitalyFedyunin) [inline test/test_torch.py:7055]: Please check format of the output

File: aten/src/ATen/native/cpu/CatKernel.cpp [hunk 2]
```diff
@@ -18,29 +17,27 @@ struct InputMeta {
...
-  for (int64_t i = 0; i < dim; i++) {
-    outer *= size[i];
-  }
-  for (int64_t i = dim + 1; i < size.size(); i++) {
-    inner *= size[i];
-  }
+void cat_contig_kernel_impl(Tensor& result, TensorList tensors, int64_t dim) {
+  int64_t outer = result.numel() / (result.size(dim) * result.stride(dim));
   scalar_t* result_data = result.data_ptr<scalar_t>();
+  // Adjust grain size to account for the work done
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/CatKernel.cpp:26]: nit: do you need this comments in the code base

File: test/test_torch.py (hunk 1 above)
Comment (VitalyFedyunin) [reply test/test_torch.py:7055]: Please add test covering both variants when `result.numel() < at::internal::GRAIN_SIZE` and when it is not. 
Please add tests covering various `dim=` values.

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
I'm concerned about changing math for `inner` and `outer`, please add tests covering various `dim=` values.

File: aten/src/ATen/native/cpu/CatKernel.cpp [hunk 3]
```diff
@@ -57,19 +54,19 @@ void cat_serial_kernel_impl(Tensor& result, TensorList tensors, int64_t dim) {
...
-      offset += local_inner;
+      result_ptr += local_inner;
     }
   }
 }
 
-void cat_serial_kernel(Tensor& result, TensorList tensors, int64_t dim) {
-  AT_DISPATCH_FLOATING_TYPES(result.scalar_type(), "cat_serial_kernel", [&]() {
-    cat_serial_kernel_impl<scalar_t>(result, tensors, dim);
+void cat_contig_kernel(Tensor& result, TensorList tensors, int64_t dim) {
```
Comment (colesbury) [inline aten/src/ATen/native/cpu/CatKernel.cpp:62]: The renaming is a bit confusing. Previously, we had: cat_serial_stub -> cat_serial_kernel -> cat_serial_kernel_impl. Everything followed the same naming pattern.

Now it is: cat_serial_stub -> cat_contig_kernel -> cat_contig_kernel_impl.

Renaming is fine if the kernel is no long single-threaded, but it looks like it still is single-threaded. Additionally, if you are going to rename the kernel, please also rename `cat_serial_stub` to match and https://github.com/pytorch/pytorch/blob/c6d301220a89908df0f346db5bf391218f8dd068/aten/src/ATen/native/cpu/CatKernel.h#L8

**Review by colesbury — COMMENTED** [review]
The approach looks reasonable, but I'm not super familiar with the `cat` code. I noticed a few issues mentioned inline.

Additionally, in the PR description you wrote:

> Enables that kernel by default for contig tensors even for large inputs and when more than one thread is available.

I don't see that change in the PR. The call to `cat_serial_stub` looks like it's still guarded by `use_serial_kernel`.

File: aten/src/ATen/native/cpu/CatKernel.cpp [hunk 4]
```diff
@@ -18,29 +17,27 @@ struct InputMeta {
     , inner_size(t.size(dim) * inner) {}
 };
 
+namespace {
```
[the remark this answers is not in the document]
Comment (kimishpatel) [reply aten/src/ATen/native/cpu/CatKernel.cpp:20]: My bad. I will fix this.

File: aten/src/ATen/native/cpu/CatKernel.cpp (hunk 2 above)
Comment (kimishpatel) [reply aten/src/ATen/native/cpu/CatKernel.cpp:26]: Yes. No need for it here.

File: aten/src/ATen/native/cpu/CatKernel.cpp (hunk 3 above)
Comment (kimishpatel) [reply aten/src/ATen/native/cpu/CatKernel.cpp:62]: My bad again. I meant to keep the name same as before, but the PR went through a series of changes and in that I half reverted my original changes. I will fix it.

File: test/test_torch.py (hunk 1 above)
Comment (kimishpatel) [reply test/test_torch.py:7055]: Sounds good. Will do.

**Comment by kimishpatel** [comment]
> The approach looks reasonable, but I'm not super familiar with the `cat` code. I noticed a few issues mentioned inline.
> 
> Additionally, in the PR description you wrote:
> 
> > Enables that kernel by default for contig tensors even for large inputs and when more than one thread is available.
> 
> I don't see that change in the PR. The call to `cat_serial_stub` looks like it's still guarded by `use_serial_kernel`.

Yes, my bad. Initially I did try what the old summary said but it did not seem to be provably better so I kept changes to the minimal. I have updated the summary.

**Comment by kimishpatel** [comment]
@VitalyFedyunin @colesbury, I have addressed your comments.

File: test/test_torch.py [hunk 5]
```diff
@@ -7045,6 +7045,56 @@ def test_cat_out_channels_last(self, device):
         res2 = torch.cat((x, y), out=z)
         self.assertEqual(res1, res2)
 
+    def test_cat_in_channels_last(self, device):
```
Comment (VitalyFedyunin) [inline test/test_torch.py:7048]: Please move test to other class where `device` is unnecessary or add cpu only decorator or use `device` appropriatly.

File: test/test_torch.py [hunk 6]
```diff
@@ -7045,6 +7045,56 @@ def test_cat_out_channels_last(self, device):
...
+        x = torch.randn((4, 15, 8, 8))
+        y = torch.randn(x.shape)
+        res1 = torch.cat((x, y), dim=0)
+        x = x.clone().contiguous(memory_format=torch.channels_last)
+        y = y.clone().contiguous(memory_format=torch.channels_last)
+        res2 = torch.cat((x, y), dim=0)
+        self.assertTrue(res2.is_contiguous(memory_format=torch.channels_last))
+        self.assertEqual(res1, res2)
+
+        # Concat across dim 2
```
Comment (VitalyFedyunin) [inline test/test_torch.py:7078]: Could be simple for loop by dimensions.

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Tests still requires some fixes.

**Comment by kimishpatel** [comment]
@VitalyFedyunin, thanks a lot of the comments. I have incorporated them. Please let me know if there is anything else.

**Review by VitalyFedyunin — APPROVED** [review]

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Windows test failures look real

**Comment by kimishpatel** [comment]
@VitalyFedyunin, looks like this PR fixes the windows failures: https://github.com/pytorch/pytorch/pull/40369.
I have rebased. Lets see.

File: aten/src/ATen/native/cpu/CatKernel.cpp [hunk 7]
```diff
@@ -57,13 +49,13 @@ void cat_serial_kernel_impl(Tensor& result, TensorList tensors, int64_t dim) {
...
       }
-      offset += local_inner;
+      result_ptr += local_inner;
     }
   }
 }
 
 void cat_serial_kernel(Tensor& result, TensorList tensors, int64_t dim) {
-  AT_DISPATCH_FLOATING_TYPES(result.scalar_type(), "cat_serial_kernel", [&]() {
+  AT_DISPATCH_FLOATING_TYPES(result.scalar_type(), "cat_contig_kernel", [&]() {
```
Comment (colesbury) [inline aten/src/ATen/native/cpu/CatKernel.cpp:58]: s/cat_contig_kernel/cat_serial_kernel

**Review by VitalyFedyunin — APPROVED** [review]
